File: tests/jobs/scala/src/main/scala/ShuffleApp.scala
import org.apache.spark.sql.SparkSession

import scala.util.Random

/**
  * This is a slight modification of GroupByTest from Spark Examples but with a deterministic
  * sequential key generation which doesn't rely on random number generator.
  *
  * Usage: ShuffleApp [numMappers] [numKeys] [valueSize] [numReducers] [sleepBeforeShutdown]
  */
object ShuffleApp {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession
      .builder
      .appName("Shuffle Test")
      .getOrCreate()

    val numMappers = if (args.length > 0) args(0).toInt else 2
    val numKeys = if (args.length > 1) args(1).toInt else 1000
    val valSize = if (args.length > 2) args(2).toInt else 1000
    val numReducers = if (args.length > 3) args(3).toInt else numMappers
    val sleepBeforeShutdownMillis = if (args.length > 4) args(4).toLong * 1000 else 0

    val keyPairs = spark.sparkContext.parallelize(0 until numMappers, numMappers).flatMap { _ =>
      val random = new Random
      val result = new Array[(Int, Array[Byte])](numKeys)

      //each partition will contain a full copy of keys to enforce shuffle
      for (i <- 0 until numKeys) {
        val byteArr = new Array[Byte](valSize)
        random.nextBytes(byteArr)
        result(i) = (i, byteArr)
      }
      result
    }.cache()
    keyPairs.count()

    println(
      s"""
         |RDD contents sample:
         | ${keyPairs.take(10).mkString(", ")}
       """.stripMargin)

    val groupsCount = keyPairs.groupByKey(numReducers).count()
    println(s"Groups count: $groupsCount")
    Thread.sleep(sleepBeforeShutdownMillis)
    spark.stop()
  }
}

File: tests/jobs/python/shuffle_app.py
from __future__ import print_function
import sys
from random import getrandbits
from time import sleep

from pyspark.sql import SparkSession

# ShuffleApp test job ported from Scala to Python

if __name__ == "__main__":
    spark = SparkSession \
        .builder \
        .appName("Shuffle Test") \
        .getOrCreate()

    num_mappers = int(sys.argv[1]) if len(sys.argv) > 1 else 2
    num_keys = int(sys.argv[2]) if len(sys.argv) > 2 else 1000
    val_size = int(sys.argv[3]) if len(sys.argv) > 3 else 1000
    num_reducers = int(sys.argv[4]) if len(sys.argv) > 4 else num_mappers
    sleep_before_shutdown_millis = int(sys.argv[5]) if len(sys.argv) > 5 else 0

    def f(_):
        result = []
        for i in range(num_keys):
            arr = bytearray(getrandbits(8) for j in range(val_size))
            result.append((i, arr))
        return result

    key_pairs = spark.sparkContext.parallelize(range(0, num_mappers), num_mappers).flatMap(f).cache()
    key_pairs.count()

    print("RDD content sample: {}".format(key_pairs.take(10)))

    groups_count = key_pairs.groupByKey(num_reducers).count()
    print("Groups count: {}".format(groups_count))

    if sleep_before_shutdown_millis > 0:
        sleep(sleep_before_shutdown_millis / 1000)

    spark.stop()
